File: analytics/src/test/java/io/appmetrica/analytics/impl/client/ClientConfigurationTest.java
package io.appmetrica.analytics.impl.client;

import android.content.Context;
import android.os.Bundle;
import io.appmetrica.analytics.BuildConfig;
import io.appmetrica.analytics.internal.CounterConfiguration;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.MockedStaticRule;
import io.appmetrica.analytics.testutils.TestUtils;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ClientConfigurationTest extends CommonTest {

    @Mock
    private CounterConfiguration counterConfiguration;
    @Mock
    private ProcessConfiguration processConfiguration;
    @Mock
    private Bundle bundle;
    private final String packageName = "packagename";
    private Context context;

    @Rule
    public MockedStaticRule<CounterConfiguration> clientConfigurationMockedStaticRule =
        new MockedStaticRule<>(CounterConfiguration.class);

    @Rule
    public MockedStaticRule<ProcessConfiguration> processConfigurationMockedStaticRule =
        new MockedStaticRule<>(ProcessConfiguration.class);

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        when(CounterConfiguration.fromBundle(bundle)).thenReturn(counterConfiguration);
        when(ProcessConfiguration.fromBundle(bundle)).thenReturn(processConfiguration);
        when(processConfiguration.getSdkApiLevel()).thenReturn(BuildConfig.API_LEVEL);
        when(processConfiguration.getPackageName()).thenReturn(packageName);
        context = TestUtils.createMockedContext();
        when(context.getPackageName()).thenReturn(packageName);
    }

    @Test
    public void testInvalidNull() {
        assertThat(ClientConfiguration.fromBundle(context, null)).isNull();
    }

    @Test
    public void testInvalidNullProcessConfiguration() {
        when(ProcessConfiguration.fromBundle(bundle)).thenReturn(null);
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNull();
    }

    @Test
    public void testInvalidDifferentPackageName() {
        when(processConfiguration.getPackageName()).thenReturn("another package name");
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNull();
    }

    @Test
    public void testInvalidDifferentSdkApiLevel() {
        when(processConfiguration.getSdkApiLevel()).thenReturn(BuildConfig.API_LEVEL - 1);
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNull();
    }

    @Test
    public void testValid() {
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNotNull();
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/client/ClientConfigurationTest.kt
package io.appmetrica.analytics.impl.client

import android.os.Bundle
import io.appmetrica.analytics.BuildConfig
import io.appmetrica.analytics.internal.CounterConfiguration
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.ContextRule
import io.appmetrica.analytics.testutils.on
import io.appmetrica.analytics.testutils.staticRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

internal class ClientConfigurationTest : CommonTest() {

    @get:Rule
    val contextRule = ContextRule()
    private val context by contextRule
    private val counterConfiguration: CounterConfiguration = mock()
    private val processConfiguration: ProcessConfiguration = mock {
        on { sdkApiLevel } doReturn BuildConfig.API_LEVEL
        on { packageName } doReturn ContextRule.PACKAGE_NAME
    }
    private val bundle: Bundle = mock()

    @get:Rule
    val clientConfigurationMockedStaticRule = staticRule<CounterConfiguration> {
        on { CounterConfiguration.fromBundle(bundle) } doReturn counterConfiguration
    }

    @get:Rule
    val processConfigurationMockedStaticRule = staticRule<ProcessConfiguration> {
        on { ProcessConfiguration.fromBundle(bundle) } doReturn processConfiguration
    }

    @Test
    fun invalidNullProcessConfiguration() {
        whenever(ProcessConfiguration.fromBundle(bundle)).thenReturn(null)
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNull()
    }

    @Test
    fun invalidDifferentPackageName() {
        whenever(processConfiguration.packageName).thenReturn("another package name")
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNull()
    }

    @Test
    fun invalidDifferentSdkApiLevel() {
        whenever(processConfiguration.sdkApiLevel).thenReturn(BuildConfig.API_LEVEL - 1)
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNull()
    }

    @Test
    fun valid() {
        assertThat(ClientConfiguration.fromBundle(context, bundle)).isNotNull()
    }
}
